vllm/v1/stats/common.py

```diff
@@ -0,0 +1,393 @@
+import time
+from dataclasses import dataclass
+from dataclasses import field as dataclass_field
+from enum import IntEnum
+from typing import List, Optional
+
+import msgspec
+from msgspec import field as msgspec_field
+
+from vllm.sampling_params import SamplingParams
+from vllm.v1.engine import EngineCoreRequest
+
+
+def ns_to_s(ns: int) -> float:
+    return ns / 1e9
+
+
+class RequestStatsUpdate(msgspec.Struct,
+                         array_like=True,
+                         omit_defaults=True,
+                         gc=False):
+    """
+    An update to the request stats.
+
+    NOTE:
+    - We should try to keep the size of this struct minimal by avoiding
+      keeping references to additional objects that are unnecessary, especially
+      when the referenced object could have been GCed already if not for
+      this reference (examples include per decoded token RequestOutput,
+      EngineCoreOutput, etc.).
+    """
+
+    class Type(IntEnum):
+        """See `RequestStats` for the lifecycle of a request."""
+        # Request arrived at the engine frontend.
+        ARRIVED = 0
+        # Input processed by the input processor.
+        INPUT_PROCESSED = 1
+        # Queued on the engine core.
+        QUEUED = 2
+        # Scheduled running by the scheduler.
+        RUNNING = 3
+        # Preempted by the scheduler.
+        PREEMPTED = 4
+        # Token decoded by the engine.
+        DECODED = 5
+        # Token detokenized by the detokenizer.
+        DETOKENIZED = 6
+
+    request_id: str
+
+    type: Type
+
+    # Timestamp when the update is recorded. This is used to record time
+    # intervals between events rather than wall clock time.
+    monotonic_ts_s: float = msgspec_field(
+        default_factory=lambda: time.monotonic())
+
+    ############################################################
+    # Metadata associated with the update.
+    ############################################################
+    # For input_processed.
+    # NOTE: it's fine to keep a reference to the engine request here
+    # because there will only be 1 event with this reference for a
+    # request. We need metadata (e.g. prompt tokens, sampling_param)
+    # from the request.
+    engine_request: Optional[EngineCoreRequest] = None
```

Comment:
I think that this has the potential to cause footguns in the future even if the initial implementation has no issues with gc. It looks like we are currently just using this for:
- `num_prompt_tokens`
- `sampling_params` 

Im not sure what the intended downstream uses of these are for the metrics, but I think we should just compute exactly what we need when this object is constructed rather than holding a reference to `EngineCoreRequest`

Reply:
Makes sense, I think having no reference with a bit more duplication might be easier to maintain in the future.